src/core/internal/hash.d

```diff
@@ -135,7 +135,7 @@ size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && (is(T ==
         static if (is(typeof(toUbyte(val)) == const(ubyte)[]))//CTFE ready for structs without reference fields
         {
             auto bytes = toUbyte(val);
-            return bytesHash(bytes.ptr, bytes.length, seed);
+            return (() @trusted => bytesHash(bytes.ptr, bytes.length, seed))();
```

Comment:
Can `bytesHash` be made safe instead?

Reply:
If it can't be made safe, and we want to use it in `@safe` context, we should try to certify it as safe and attribute the entire function as @trusted.

Reply:
Perhaps add an overload to take an array instead of pointer and length pair as well.

Reply:
It's doing pointer arithmetic in a new places, it's explicitly marked as `@system`.

Reply:
>Perhaps add an overload to take an array instead of pointer and length pair as well.

I could add a `@trusted` overload that calls `bytesHash` if you think that buys us anything over the `@trusted` lambda that calls `bytesHash`. My sole concern is that it might result in worse machine code, since the compiler might treat inline lambdas more intelligently. I'll take a look.

Reply:
This PR appears to be an attempt to silence the compiler rather than heeding its advice.  Essentially this PR is just casting away `byteHash`'s unsafeness.

When we attribute anything as `@trusted`, we are saying to our users that we have thoroughly analyzed the implementation and certified it as memory safe, even though the implementation utilizes unsafe language features.  I'm not seeing any evidence that that due diligence has been done.

IMO, we need to rewrite `bytesHash` using the language's memory-safe guarantees and attribute it with `@safe`, or we need to do a thorough analysis of the `bytesHash` function to certify it as memory-safe, ad attribute it with `@trusted`.

Reply:
>This PR appears to be an attempt to silence the compiler rather than heeding its advice. Essentially this PR is just casting away byteHash's unsafeness.

Yes. Although the function is unsafe in general due to the lack of language-level enforcement that the first two arguments describe a valid array slice, we know the call is safe here because we are passing it the pointer and length of an actual array. I believe this is an appropriate use of `@trusted`.

Reply:
@n8sh Ok.  I'll need to look a little closer at this PR then, but at a minimum that explanation should accompany the `@trusted` attribution in the form of a comment.